python_training/XO.py

```diff
@@ -0,0 +1,181 @@
+        if self._check_for_winning_sequence(left_to_right_diagonal):
+            return True
+
+        right_to_left_diagonal = [self.board[self.board_size - 1 - index][index] for index in range(self.board_size)]
+        if self._check_for_winning_sequence(right_to_left_diagonal):
+            return True
+
+        if self.turns_counter == self.board_size ** 2:
+            print("The game has ended with a tie!")
+            return True
+
+        return False
+
+    def _check_for_winning_sequence(self, sequence) -> bool:
+        """
+        Check if a specific sequence is comprised of a uniform symbol, and indicates a win.
+
+        :param sequence: The sequence that should be checked.
+        """
+        if sequence.count(self.current_player) == self.board_size:
+            print('The winner is {}'.format(self.players[self.current_player]))
+            return True
+
+        return False
+
+    def _get_validated_slot_number_input(self):
+        """
+        Get a valid slot number input from the user.
+
+        :return: a tuple with two items: (row index, column index)
+        """
+        is_slot_number_valid = False
+
+        while not is_slot_number_valid:
+            row_index = input("Please enter row number >")
+            column_index = input("Please enter column number >")
+            if not row_index.isnumeric() or not column_index.isnumeric():
+                print("Slot locations must be numbers, please try again")
+                continue
+            row_index = int(row_index)
+            column_index = int(column_index)
+            if row_index not in range(self.board_size) or column_index not in range(self.board_size):
+                print("Make sure the slot numbers you have entered are within the range of the board.")
+            elif self.board[row_index][column_index]:
+                print("That slot is already taken, choose again")
+            else:
+                is_slot_number_valid = True
+
+        return row_index, column_index
+
+
+def get_validated_board_size() -> int:
+    """
+    Validates that the user has given a numeric board size that is bigger than 2.
+
+    :return: A board size after validation.
+    """
+    is_board_size_valid = False
+    while not is_board_size_valid:
+        board_size_input = input('Please enter board size')
+        if not board_size_input.isnumeric():
+            print("invalid input, please enter a number")
+            continue
+        board_size_input = int(board_size_input)
+        if board_size_input < 3:
+            print('board size must be bigger than 3')
+        else:
+            is_board_size_valid = True
+
+    return board_size_input
+
+
+def main():
+    player1 = input('Please enter the name of player one >')
+    player2 = input('Please enter the name of player two >')
+    board_size = get_validated_board_size()
+
+    game = XOGame(player1, player2, board_size)
+    game.start_game()
+
+
+if __name__ == "__main__":
```

Comment:
Always use one `'` instead of `"` (not important at all, but just for convention).